diffy/plugins/diffy_osquery/plugin.py

```diff
@@ -32,30 +33,41 @@ def generate(self, incident: str, **kwargs) -> List[str]:
         logger.debug("Generating osquery payload.")
         session = Session()
 
-        # TODO check for existence before deployment
-        # we run with these commands with diffy credentials so as to not pollute the on-instance credentials
-        creds = session.get_credentials()
-        region = kwargs.get("region", CONFIG.get("DIFFY_PAYLOAD_OSQUERY_REGION"))
-        key = kwargs.get("key", CONFIG.get("DIFFY_PAYLOAD_OSQUERY_KEY"))
-
-        if not region:
-            raise BadArguments(
-                "DIFFY_PAYLOAD_OSQUERY_REGION required for use with OSQuery plugin."
-            )
-
-        if not key:
-            raise BadArguments(
-                "DIFFY_PAYLOAD_OSQUERY_KEY required for use with OSQuery plugin."
-            )
-
-        commands: List[str] = [
-            f"export AWS_ACCESS_KEY_ID={creds.access_key}",
-            f"export AWS_SECRET_ACCESS_KEY={creds.secret_key}",
-            f"export AWS_SESSION_TOKEN={creds.token}",
-            f"cd $(mktemp -d -t binaries-{incident}-`date +%s`-XXXXXX)",
-            f"aws s3 --region {region} cp s3://{key} ./latest.tar.bz2 --quiet",
-            "tar xvf latest.tar.bz2 &>/dev/null",
-        ]
+        # If osquery isn't present, obtain an osquery binary from S3.
+        if not which("osqueryi"):
+            # We run these commands with Diffy credentials so as to not pollute
+            # the on-instance credentials.
+            creds = session.get_credentials()
+            region = kwargs.get("region", CONFIG.get("DIFFY_PAYLOAD_OSQUERY_REGION"))
+            key = kwargs.get("key", CONFIG.get("DIFFY_PAYLOAD_OSQUERY_KEY"))
+
+            if not region:
+                raise BadArguments(
+                    "DIFFY_PAYLOAD_OSQUERY_REGION required for use with OSQuery plugin."
+                )
+
+            if not key:
+                raise BadArguments(
+                    "DIFFY_PAYLOAD_OSQUERY_KEY required for use with OSQuery plugin."
+                )
+
+            # If we've downloaded our own osquery collection binary, create a
+            # symbolic link, allowing us to use relative commands elsewhere.
+            commands: List[str] = [
+                f"export AWS_ACCESS_KEY_ID={creds.access_key}",
+                f"export AWS_SECRET_ACCESS_KEY={creds.secret_key}",
+                f"export AWS_SESSION_TOKEN={creds.token}",
+                f"cd $(mktemp -d -t binaries-{incident}-`date +%s`-XXXXXX)",
+                f"aws s3 --region {region} cp s3://{key} ./latest.tar.bz2 --quiet",
+                "tar xvf latest.tar.bz2 &>/dev/null",
+                "export PATH=${PATH}:${HOME}/.local/bin",
+                "mkdir -p ${HOME}/.local/bin",
+                "ln -s ./usr/bin/osqueryi ${HOME}/.local/bin/osqueryi",
```

Comment:
Should we make an effort to clean up the osquery binary like we would the rest of the incident binary's?

Reply:
I was thinking about it, and unless it's "forensic purity," I'm not really sure what the justification for cleanup is. But even _that_ is a whole set of modifications to system state probably better left undone, if we've already gone so far as to download software to the machine. It certainly doesn't make us any faster. We could consider running everything only in memory?

Reply:
I think that's reasonable, do we need to audit the other payloads as well? If nothing else I think we should try to be consistent.

Reply:
Yeah, agreed. We should be good re: cleanup.